File: base-module/src/main/java/com/ezteam/baseproject/dialog/GuideEditDialog.java
package com.ezteam.baseproject.dialog;

import android.app.Dialog;
import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;

import androidx.core.content.ContextCompat;
import androidx.core.view.WindowCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.core.view.WindowInsetsControllerCompat;

import android.widget.TextView;

import com.ezteam.baseproject.R;

import java.util.List;

public class GuideEditDialog extends Dialog {
    private int currentStepIndex = 0;
    private final List<GuideStep> steps;
    private final View rootView;
    private final SpotlightOverlayView spotlightOverlay;
    private final ImageView arrowImage;
    private final LinearLayout contentBox;
    private final Button btnGotIt;
    private final Context context;
    private OnStepChangedListener onStepChangedListener;
    private final int borderColor;

    public GuideEditDialog(Context context, List<GuideStep> steps, int borderColor) {
        super(context);
        this.context = context;
        this.steps = steps;
        this.borderColor = borderColor;

        LayoutInflater inflater = LayoutInflater.from(context);
        rootView = inflater.inflate(R.layout.guide_edit_spotlight, null);
        setContentView(rootView);

        if (getWindow() != null) {
            getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
            getWindow().setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
            WindowInsetsControllerCompat insetsController = WindowCompat.getInsetsController(getWindow(), getWindow().getDecorView());
            if (insetsController != null) {
                insetsController.setSystemBarsBehavior(WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE);
                insetsController.hide(WindowInsetsCompat.Type.navigationBars());
            }
        }

        spotlightOverlay = (SpotlightOverlayView) rootView.findViewById(R.id.spotlight_overlay);
        contentBox = rootView.findViewById(R.id.content_box);
        btnGotIt = rootView.findViewById(R.id.btn_got_it);
        btnGotIt.getBackground().setTint(borderColor);
        arrowImage = rootView.findViewById(R.id.arrow_image);
       // arrowImage.setColorFilter(borderColor);

        showStep(currentStepIndex);

        btnGotIt.setOnClickListener(v -> {
            currentStepIndex++;
           // spotlightOverlay.stopBlinking();
            if (currentStepIndex < steps.size()) {
                showStep(currentStepIndex);
            } else {
                dismiss();
            }
        });
    }

    private void showStep(int index) {
        if (index < 0 || index >= steps.size()) return;

        GuideStep step = steps.get(index);

        if (onStepChangedListener != null) {
            onStepChangedListener.onStepChanged(index);
        }
        try {
            step.targetView.post(() -> {
                int[] loc = new int[2];
                step.targetView.getLocationOnScreen(loc);
                int targetX = loc[0];
                int targetY = loc[1];
                int targetWidth = step.targetView.getWidth();
                int targetHeight = step.targetView.getHeight();
                spotlightOverlay.setHoleAroundView(step.targetView, 16);
//                spotlightOverlay.startBlinkingBorder(borderColor,1000L);

                int[] rootLoc = new int[2];
                rootView.getLocationOnScreen(rootLoc);

                int arrowWidth = arrowImage.getWidth();
                int arrowHeight = arrowImage.getHeight();

                int arrowX = targetX - rootLoc[0] + targetWidth / 2 - arrowWidth / 2;
                int arrowY = targetY - rootLoc[1] - arrowHeight - 16 + (int) step.arrowOffsetY;

                arrowX = Math.max(0, Math.min(arrowX, rootView.getWidth() - arrowWidth));
                arrowY = Math.max(0, Math.min(arrowY, rootView.getHeight() - arrowHeight));

                arrowImage.setX(arrowX);
                arrowImage.setY(arrowY);

                updateArrowRotation(step.step);
                arrowImage.setVisibility(View.VISIBLE);

                int childCount = contentBox.getChildCount();
                if (childCount > 1) {
                    contentBox.removeViews(0, childCount - 1);
                }

                for (String text : step.titleLines) {
                    TextView tv = new TextView(context);
                    tv.setText(text);
                    tv.setTextColor(ContextCompat.getColor(context, R.color.text1));
                    tv.setTextSize(14f);
                    tv.setPadding(0, 4, 0, 0);
                    tv.setSingleLine(false);
                    tv.setMaxLines(Integer.MAX_VALUE);
                    tv.setEllipsize(null);
                    tv.setLayoutParams(new LinearLayout.LayoutParams(
                            ViewGroup.LayoutParams.WRAP_CONTENT,
                            ViewGroup.LayoutParams.WRAP_CONTENT
                    ));
                    contentBox.addView(tv, contentBox.getChildCount() - 1);
                }
            });
        } catch (Exception e){
            Log.e("GuideEditDialog", "error " + e);
        }

    }

    private void updateArrowRotation(int index) {
        switch (index) {
            case 0:
                arrowImage.setRotation(180);
                break;
            case 1:
                arrowImage.setRotation(90);
                break;
            case 2:
                arrowImage.setRotation(270);
                break;
            default:
                arrowImage.setRotation(0);
        }
    }

    public interface OnStepChangedListener {
        void onStepChanged(int stepIndex);
    }

    public void setOnStepChangedListener(OnStepChangedListener listener) {
        this.onStepChangedListener = listener;
    }

    public void setOnDismissListener(OnDismissListener listener) {
        super.setOnDismissListener(dialog -> {
            btnGotIt.getBackground().setTint(context.getResources().getColor(R.color.primaryColor));
            if (listener != null) listener.onDismiss(dialog);
        });
    }
}

File: base-pdf-reader/src/main/java/com/ezteam/ezpdflib/guide/GuideEditDialog.kt
package com.ezteam.ezpdflib.guide

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.animation.ValueAnimator
import android.app.Dialog
import android.content.Context
import android.graphics.*
import android.graphics.drawable.ColorDrawable
import android.os.Handler
import android.os.Looper
import android.text.TextUtils
import android.view.*
import android.view.animation.AlphaAnimation
import android.view.animation.TranslateAnimation
import android.widget.Button
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat
import com.ezteam.baseproject.databinding.GuideEditSpotlightBinding
import com.ezteam.ezpdflib.R

class GuideEditDialog(
    context: Context,
    private val steps: List<GuideStep>
) : Dialog(context) {

    private lateinit var binding: GuideEditSpotlightBinding
    private var currentStepIndex = 0

    init {
        binding = GuideEditSpotlightBinding.inflate(LayoutInflater.from(context))
        setContentView(binding.root)
        val insetsController = WindowCompat.getInsetsController(window!!, window!!.decorView)
        insetsController.systemBarsBehavior =
            WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        insetsController.hide(WindowInsetsCompat.Type.navigationBars())
        window?.apply {
            setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
            setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        }

        showStep(currentStepIndex)

        binding.btnGotIt.setOnClickListener {
            currentStepIndex++
            if (currentStepIndex < steps.size) {
                showStep(currentStepIndex)
            } else {
                dismiss()
                stopArrowDiagonal(binding.arrowImage)
            }
        }
    }

    private var arrowDiagonalAnim: AnimatorSet? = null
    private var shouldLoopArrow = false

    private fun startArrowDiagonal(
        arrow: ImageView,
        distanceDp: Float = 160f,
        startOffsetXDp: Float = 0f,
        startOffsetYDp: Float = 0f
    ) {
        val density = arrow.resources.displayMetrics.density
        val d = distanceDp * density
        val startXOffset = startOffsetXDp * density
        val startYOffset = startOffsetYDp * density

        shouldLoopArrow = true

        arrow.translationX = startXOffset
        arrow.translationY = startYOffset
        arrow.alpha = 0f
        arrow.visibility = View.VISIBLE

        val fadeIn = ObjectAnimator.ofFloat(arrow, View.ALPHA, 0f, 1f).apply { duration = 150 }
        val moveX =
            ObjectAnimator.ofFloat(arrow, View.TRANSLATION_X, startXOffset, startXOffset + d)
        val moveY =
            ObjectAnimator.ofFloat(arrow, View.TRANSLATION_Y, startYOffset, startYOffset + d)
        val move = AnimatorSet().apply {
            playTogether(moveX, moveY)
            duration = 700
        }
        val fadeOut = ObjectAnimator.ofFloat(arrow, View.ALPHA, 1f, 0f).apply { duration = 200 }

        arrowDiagonalAnim?.cancel()
        arrowDiagonalAnim = AnimatorSet().apply {
            playSequentially(fadeIn, move, fadeOut)
            addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationEnd(animation: Animator) {
                    // Chỉ lặp khi còn ở bước cần hiệu ứng
                    if (isShowing && shouldLoopArrow) {
                        arrow.postDelayed({
                            if (shouldLoopArrow) {
                                startArrowDiagonal(
                                    arrow,
                                    distanceDp,
                                    startOffsetXDp,
                                    startOffsetYDp
                                )
                            }
                        }, 300)
                    }
                }
            })
            start()
        }
    }

    private fun stopArrowDiagonal(arrow: ImageView) {
        shouldLoopArrow = false
        arrowDiagonalAnim?.removeAllListeners()
        arrowDiagonalAnim?.cancel()
        arrow.clearAnimation()
        arrow.animate().cancel()
        arrow.alpha = 1f
    }

    private fun dp(dp: Float, v: View): Float = dp * v.resources.displayMetrics.density

    private fun getRelativeTopLeftOnOverlay(child: View, overlay: View): PointF {
        val childLoc = IntArray(2)
        val overlayLoc = IntArray(2)
        child.getLocationOnScreen(childLoc)
        overlay.getLocationOnScreen(overlayLoc)
        val relX = (childLoc[0] - overlayLoc[0]).toFloat()
        val relY = (childLoc[1] - overlayLoc[1]).toFloat()
        return PointF(relX, relY)
    }

    private fun showStep(index: Int) {
        val step = steps[index]
        step.targetView.post {
            binding.spotlightOverlay.setHoleAroundView(step.targetView, margin = 12)
//            binding.spotlightOverlay.startBlinkingBorder()

            val arrow = binding.arrowImage

            val topLeft = getRelativeTopLeftOnOverlay(step.targetView, binding.spotlightOverlay)
            val targetCenterX = topLeft.x + step.targetView.width / 2f
            val targetTopY    = topLeft.y
            val targetCenterY = topLeft.y + step.targetView.height / 2f

            val off = dp(24f, arrow)

            arrow.post {
                arrow.x = targetCenterX - arrow.width / 2f - off + dp(20f, arrow)
                arrow.y = targetTopY - off + step.arrowOffsetY

                if (index == 0) {
                    arrow.x = topLeft.x
                    arrow.y = topLeft.y
                    startArrowDiagonal(
                        arrow,
                        distanceDp = 80f,
                        startOffsetXDp = -220f,
                        startOffsetYDp = -50f
                    )
                } else {
                    stopArrowDiagonal(arrow)
                }

                val tvGuide = (binding.root as ViewGroup).findViewById<TextView>(R.id.tv_guide_text)
                tvGuide.text = step.titleLines.joinToString("\n")
                tvGuide.setTextColor(ContextCompat.getColor(context, R.color.text1))
                tvGuide.textSize = 14f
                tvGuide.ellipsize = null
                tvGuide.isSingleLine = false
            }
        }
    }
}

